File: BFS/Max_Depth_of_N-ary_Tree.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    

public:
    int maxDepth(Node* root) {
        if(!root)
            return 0;
        int depth = 0;
       queue<Node*> q;
        q.push(root);
        while(!q.empty())
        {
            depth +=1;
           int len = q.size();
            for(int i=0;i<len;i++)
            {
                Node* node = q.front();
                q.pop();
                 for(auto child : node->children)
                 {
                    if(child)
                        q.push(child);
                }
            }
           
        }
        return depth;
        
    }
};



/* DFS */
class Solution {
    

public:
    int maxDepth(Node* root) {
        if(!root)
            return 0;
        
        int depth = 0;
        for(auto child : root->children)
        {
            depth = max(depth,maxDepth(child));
        }
        return depth+1;
        
    }
};

File: BFS/Java/Easy/Max_Depth_of_N-Ary_Tree.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    public int maxDepth(Node root) {
        if(root == null)
            return 0;
        Queue<Node> queue = new LinkedList<Node>();
        int depth = 0;
        queue.add(root);
        while(!queue.isEmpty()){
            int size = queue.size();
            for(int i=0;i<size;i++)
            {
                Node currNode = queue.poll();
                queue.addAll(currNode.children);
            }
            depth++;
            
        }
        return depth;
    }
}